experimental/wbuild-examples/FWF/src/Icon/Icon.man
'\" t
.TH "" 3 "" "Version 3.0" "Free Widget Foundation"
.SH NAME
XfwfIcon
.SH DESCRIPTION
The XfwfIcon widget displays an image. The preferred width and
height of the widget will be set to the width and height of the loaded
image, plus space for the frame. The widget has a callback
\fIactivateCallback\fP, that is by default bound to a click of the left
mouse button.

The image is a \fIPixmap\fP, optionally accompanied by a mask, to make
parts of the image transparent. When the image has a mask, the frame
around the widget will not be shown.

.SS "Public variables"

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfIcon
Name	Class	Type	Default
XtNimage	XtCImage	Icon *	NULL 
XtNactivate	XtCActivate	Callback	NULL 

.TE
.ps +2

.TP
.I "XtNimage"
The image must be in \fIPixmap\fP format. The width and height of the
image will be used to set the width and height of the widget.  There
is a converter for strings (defined in the Common widget), which will
try to interpret a string as a file name and load the file. The file
must be in XPM format. The converter also knows about some built-in
images, called \fI"FATAL"\fP, \fI"ERROR"\fP, \fI"WARNING"\fP, \fI"QUESTION"\fP,
\fI"INFO"\fP and \fI"NONE"\fP.

	

.TP
.I "XtNactivate"
The callback is called by the \fIactivate\fP action, which is by default
bound to a click of mouse button 1.

	

.TP
.I "XtNframeWidth"
By default, icons do not have a frame.

	

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfBoard
Name	Class	Type	Default
XtNabs_x	XtCAbs_x	Position 	0 
XtNrel_x	XtCRel_x	Float 	"0.0"
XtNabs_y	XtCAbs_y	Position 	0 
XtNrel_y	XtCRel_y	Float 	"0.0"
XtNabs_width	XtCAbs_width	Position 	0 
XtNrel_width	XtCRel_width	Float 	"1.0"
XtNabs_height	XtCAbs_height	Position 	0 
XtNrel_height	XtCRel_height	Float 	"1.0"
XtNhunit	XtCHunit	Float 	"1.0"
XtNvunit	XtCVunit	Float 	"1.0"
XtNlocation	XtCLocation	String 	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfFrame
Name	Class	Type	Default
XtNcursor	XtCCursor	Cursor 	None 
XtNframeType	XtCFrameType	FrameType 	XfwfRaised 
XtNframeWidth	XtCFrameWidth	Dimension 	0 
XtNouterOffset	XtCOuterOffset	Dimension 	0 
XtNinnerOffset	XtCInnerOffset	Dimension 	0 
XtNshadowScheme	XtCShadowScheme	ShadowScheme 	XfwfAuto 
XtNtopShadowColor	XtCTopShadowColor	Color 	compute_topcolor 
XtNbottomShadowColor	XtCBottomShadowColor	Color 	compute_bottomcolor 
XtNtopShadowStipple	XtCTopShadowStipple	Bitmap 	NULL 
XtNbottomShadowStipple	XtCBottomShadowStipple	Bitmap 	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
XfwfCommon
Name	Class	Type	Default
XtNuseXCC	XtCUseXCC	Boolean 	TRUE 
XtNusePrivateColormap	XtCUsePrivateColormap	Boolean 	FALSE 
XtNuseStandardColormaps	XtCUseStandardColormaps	Boolean 	TRUE 
XtNstandardColormap	XtCStandardColormap	Atom 	0 
XtNxcc	XtCXCc	XCC 	create_xcc 
XtNtraversalOn	XtCTraversalOn	Boolean 	True 
XtNhighlightThickness	XtCHighlightThickness	Dimension 	2 
XtNhighlightColor	XtCHighlightColor	Color 	XtDefaultForeground 
XtNbackground	XtCBackground	Color 	XtDefaultBackground 
XtNhighlightPixmap	XtCHighlightPixmap	Pixmap 	None 
XtNnextTop	XtCNextTop	Callback	NULL 
XtNuserData	XtCUserData	Pointer	NULL 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Composite
Name	Class	Type	Default
XtNchildren	XtCChildren	WidgetList 	NULL 
insertPosition	XtCInsertPosition	XTOrderProc 	NULL 
numChildren	XtCNumChildren	Cardinal 	0 

.TE
.ps +2

.ps -2
.TS
center box;
cBsss
lB|lB|lB|lB
l|l|l|l.
Core
Name	Class	Type	Default
XtNx	XtCX	Position 	0 
XtNy	XtCY	Position 	0 
XtNwidth	XtCWidth	Dimension 	0 
XtNheight	XtCHeight	Dimension 	0 
borderWidth	XtCBorderWidth	Dimension 	0 
XtNcolormap	XtCColormap	Colormap 	NULL 
XtNdepth	XtCDepth	Int 	0 
destroyCallback	XtCDestroyCallback	XTCallbackList 	NULL 
XtNsensitive	XtCSensitive	Boolean 	True 
XtNtm	XtCTm	XTTMRec 	NULL 
ancestorSensitive	XtCAncestorSensitive	Boolean 	False 
accelerators	XtCAccelerators	XTTranslations 	NULL 
borderColor	XtCBorderColor	Pixel 	0 
borderPixmap	XtCBorderPixmap	Pixmap 	NULL 
background	XtCBackground	Pixel 	0 
backgroundPixmap	XtCBackgroundPixmap	Pixmap 	NULL 
mappedWhenManaged	XtCMappedWhenManaged	Boolean 	True 
XtNscreen	XtCScreen	Screen *	NULL 

.TE
.ps +2

.SS "Exports"

The type \fIIcon\fP is defined in \fIConverters.h\fP

.nf

.B incl
 <Xfwf/Converters.h>
.fi

.SS "Translations"

By default, the \fIactivate\fP action is bound to a mouse click and to
the Return key.

	

.nf
<Btn1Down>,<Btn1Up>: activate() 
.fi

.nf
<Key>Return: activate() 
.fi

.SS "Actions"

.TP
.I "activate

The \fIactivate\fP action just calls the \fIactivate\fP callback functions,
passing the \fIXEvent\fP pointer in the \fIcall_data\fP argument.



experimental/wbuild-examples/FWF/src/Icon/Icon.c
/* Generated by wbuild from "Icon.w"
** (generator version $Revision: 3.0 $ of $Date: 95/08/30 17:50:38 $)
*/
#include <X11/IntrinsicP.h>
#include <X11/StringDefs.h>
#line 193 "Icon.w"
#include <X11/extensions/shape.h>
#line 194 "Icon.w"
#include <stdio.h>
#include <Xfwf/IconP.h>
#line 161 "Icon.w"
static void activate(
#if NeedFunctionPrototypes
Widget,XEvent*,String*,Cardinal*
#endif
);

static XtActionsRec actionsList[] = {
{"activate", activate},
};

static char defaultTranslations[] = "\
<Btn1Down>,<Btn1Up>: activate() \n\
<Key>Return: activate() \n\
";
static void _resolve_inheritance(
#if NeedFunctionPrototypes
WidgetClass
#endif
);
#line 57 "Icon.w"
static void initialize(
#if NeedFunctionPrototypes
Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 85 "Icon.w"
static void expose(
#if NeedFunctionPrototypes
Widget,XEvent *,Region 
#endif
);
#line 101 "Icon.w"
static Boolean  set_values(
#if NeedFunctionPrototypes
Widget ,Widget ,Widget,ArgList ,Cardinal *
#endif
);
#line 136 "Icon.w"
static void realize(
#if NeedFunctionPrototypes
Widget,XtValueMask *,XSetWindowAttributes *
#endif
);
#line 172 "Icon.w"
static void create_image_gc(
#if NeedFunctionPrototypes
Widget
#endif
);
#line 172 "Icon.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 172 "Icon.w"
static void create_image_gc(Widget self)
#else
#line 172 "Icon.w"
static void create_image_gc(self)Widget self;
#endif
#line 173 "Icon.w"
{
    Dimension wd, ht;
    Position x, y;
    XtGCMask mask = GCFillStyle | GCTileStipXOrigin | GCTileStipYOrigin;
    XGCValues values;

    if (((XfwfIconWidget)self)->xfwfIcon.image_gc != NULL) XtReleaseGC(self, ((XfwfIconWidget)self)->xfwfIcon.image_gc);
    ((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &wd, &ht);
    if (((XfwfIconWidget)self)->xfwfIcon.image && ((XfwfIconWidget)self)->xfwfIcon.image->pixmap != None) {
	values.tile = ((XfwfIconWidget)self)->xfwfIcon.image->pixmap;
	mask |= GCTile;
    }
    values.fill_style = FillTiled;
    values.ts_x_origin = x;
    values.ts_y_origin = y;
    ((XfwfIconWidget)self)->xfwfIcon.image_gc = XtGetGC(self, mask, &values);
}

static XtResource resources[] = {
#line 28 "Icon.w"
{XtNimage,XtCImage,XtRIcon,sizeof(((XfwfIconRec*)NULL)->xfwfIcon.image),XtOffsetOf(XfwfIconRec,xfwfIcon.image),XtRImmediate,(XtPointer)NULL },
#line 33 "Icon.w"
{XtNactivate,XtCActivate,XtRCallback,sizeof(((XfwfIconRec*)NULL)->xfwfIcon.activate),XtOffsetOf(XfwfIconRec,xfwfIcon.activate),XtRImmediate,(XtPointer)NULL },
#line 37 "Icon.w"
{XtNframeWidth,XtCFrameWidth,XtRDimension,sizeof(((XfwfIconRec*)NULL)->xfwfFrame.frameWidth),XtOffsetOf(XfwfIconRec,xfwfFrame.frameWidth),XtRImmediate,(XtPointer)0 },
};

XfwfIconClassRec xfwfIconClassRec = {
{ /* core_class part */
/* superclass   	*/  (WidgetClass) &xfwfBoardClassRec,
/* class_name   	*/  "XfwfIcon",
/* widget_size  	*/  sizeof(XfwfIconRec),
/* class_initialize 	*/  NULL,
/* class_part_initialize*/  _resolve_inheritance,
/* class_inited 	*/  FALSE,
/* initialize   	*/  initialize,
/* initialize_hook 	*/  NULL,
/* realize      	*/  realize,
/* actions      	*/  actionsList,
/* num_actions  	*/  1,
/* resources    	*/  resources,
/* num_resources 	*/  3,
/* xrm_class    	*/  NULLQUARK,
/* compres_motion 	*/  True ,
/* compress_exposure 	*/  XtExposeCompressMultiple ,
/* compress_enterleave 	*/  True ,
/* visible_interest 	*/  TRUE ,
/* destroy      	*/  NULL,
/* resize       	*/  XtInheritResize,
/* expose       	*/  expose,
/* set_values   	*/  set_values,
/* set_values_hook 	*/  NULL,
/* set_values_almost 	*/  XtInheritSetValuesAlmost,
/* get_values+hook 	*/  NULL,
/* accept_focus 	*/  XtInheritAcceptFocus,
/* version      	*/  XtVersion,
/* callback_private 	*/  NULL,
/* tm_table      	*/  defaultTranslations,
/* query_geometry 	*/  XtInheritQueryGeometry,
/* display_acceleator 	*/  XtInheritDisplayAccelerator,
/* extension    	*/  NULL 
},
{ /* composite_class part */
XtInheritGeometryManager,
XtInheritChangeManaged,
XtInheritInsertChild,
XtInheritDeleteChild,
NULL
},
{ /* XfwfCommon_class part */
XtInherit_compute_inside,
XtInherit_total_frame_width,
XtInherit_highlight_border,
XtInherit_unhighlight_border,
XtInherit_would_accept_focus,
XtInherit_traverse,
XtInherit_lighter_color,
XtInherit_darker_color,
/* traversal_trans */  NULL ,
},
{ /* XfwfFrame_class part */
 /* dummy */  0
},
{ /* XfwfBoard_class part */
XtInherit_set_abs_location,
},
{ /* XfwfIcon_class part */
 /* dummy */  0
},
};
WidgetClass xfwfIconWidgetClass = (WidgetClass) &xfwfIconClassRec;
/*ARGSUSED*/
#line 161 "Icon.w"
static void activate(self,event,params,num_params)Widget self;XEvent*event;String*params;Cardinal*num_params;
{
    XtCallCallbackList(self, ((XfwfIconWidget)self)->xfwfIcon.activate, event);
}

static void _resolve_inheritance(class)
WidgetClass class;
{
  XfwfIconWidgetClass c = (XfwfIconWidgetClass) class;
  XfwfIconWidgetClass super;
  static CompositeClassExtensionRec extension_rec = {
    NULL, NULLQUARK, XtCompositeExtensionVersion,
    sizeof(CompositeClassExtensionRec), True};
  CompositeClassExtensionRec *ext;
  ext = (XtPointer)XtMalloc(sizeof(*ext));
  *ext = extension_rec;
  ext->next_extension = c->composite_class.extension;
  c->composite_class.extension = ext;
  if (class == xfwfIconWidgetClass) return;
  super = (XfwfIconWidgetClass)class->core_class.superclass;
}
#line 57 "Icon.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 57 "Icon.w"
static void initialize(Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 57 "Icon.w"
static void initialize(request,self,args,num_args)Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 58 "Icon.w"
{
    Dimension dummy1, dummy2;
    Position x, y;

    ((XfwfIconWidget)self)->xfwfIcon.image_gc = NULL;
    create_image_gc(self);
    if (((XfwfIconWidget)self)->xfwfIcon.image && ((XfwfIconWidget)self)->xfwfIcon.image->pixmap != None) {
	((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &dummy1, &dummy2);
	((XfwfIconWidget)self)->core.width = ((XfwfIconWidget)self)->xfwfIcon.image->attributes.width + 2*x;
	((XfwfIconWidget)self)->core.height = ((XfwfIconWidget)self)->xfwfIcon.image->attributes.height + 2*y;
	/* fprintf(stderr, "Icon.initialize: width=%d height=%d\n",
		$width, $height); */
    } else {
	((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &dummy1, &dummy2);
	((XfwfIconWidget)self)->core.width = 10 + 2*x;
	((XfwfIconWidget)self)->core.height = 10 + 2*y;
	/* fprintf(stderr, "Icon.initialize: width=%d height=%d (no icon)\n",
		$width, $height); */
    }
}
#line 85 "Icon.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 85 "Icon.w"
static void expose(Widget self,XEvent * event,Region  region)
#else
#line 85 "Icon.w"
static void expose(self,event,region)Widget self;XEvent * event;Region  region;
#endif
#line 86 "Icon.w"
{
    Dimension wd, ht;
    Position x, y;

    if (! XtIsRealized(self)) return;
    if (((XfwfIconWidget)self)->xfwfIcon.image && ((XfwfIconWidget)self)->xfwfIcon.image->pixmap != None) {
	((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &wd, &ht);
	XFillRectangle(XtDisplay(self), XtWindow(self), ((XfwfIconWidget)self)->xfwfIcon.image_gc, x, y, wd, ht);
    }
    xfwfBoardClassRec.core_class.expose(self, event, region);
}
#line 101 "Icon.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 101 "Icon.w"
static Boolean  set_values(Widget  old,Widget  request,Widget self,ArgList  args,Cardinal * num_args)
#else
#line 101 "Icon.w"
static Boolean  set_values(old,request,self,args,num_args)Widget  old;Widget  request;Widget self;ArgList  args;Cardinal * num_args;
#endif
#line 102 "Icon.w"
{
    Dimension dummy1, dummy2;
    Position x, y;
    Boolean need_redraw = False;

    if (((XfwfIconWidget)old)->xfwfFrame.frameWidth != ((XfwfIconWidget)self)->xfwfFrame.frameWidth
	|| ((XfwfIconWidget)old)->xfwfFrame.outerOffset != ((XfwfIconWidget)self)->xfwfFrame.outerOffset
	|| ((XfwfIconWidget)old)->xfwfIcon.image != ((XfwfIconWidget)self)->xfwfIcon.image) {
	create_image_gc(self);
	need_redraw = True;
    }
    if (((XfwfIconWidget)old)->xfwfIcon.image != ((XfwfIconWidget)self)->xfwfIcon.image) {
	if (! ((XfwfIconWidget)self)->xfwfIcon.image) {				/* Remove shape mask */
	    XShapeCombineMask(XtDisplay(self), XtWindow(self), ShapeBounding,
			      0, 0, None, ShapeSet);
	} else {				/* New size and shape */
	    ((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &dummy1, &dummy2);
	    if (((XfwfIconWidget)self)->xfwfIcon.image->pixmap != None
		&& (((XfwfIconWidget)self)->xfwfIcon.image->attributes.valuemask & XpmSize) != 0)
		((XfwfIconWidgetClass)self->core.widget_class)->xfwfBoard_class.set_abs_location(self, CWWidth | CWHeight, 0, 0,
				  ((XfwfIconWidget)self)->xfwfIcon.image->attributes.width + 2*x,
				  ((XfwfIconWidget)self)->xfwfIcon.image->attributes.height + 2*y);
	    if (XtIsRealized(self))
		XShapeCombineMask(XtDisplay(self), XtWindow(self), ShapeBounding,
				  x, y, ((XfwfIconWidget)self)->xfwfIcon.image->mask, ShapeSet);
	}
	need_redraw = True;
    }
    return need_redraw;
}
#line 136 "Icon.w"
/*ARGSUSED*/
#if NeedFunctionPrototypes
#line 136 "Icon.w"
static void realize(Widget self,XtValueMask * mask,XSetWindowAttributes * attributes)
#else
#line 136 "Icon.w"
static void realize(self,mask,attributes)Widget self;XtValueMask * mask;XSetWindowAttributes * attributes;
#endif
#line 137 "Icon.w"
{
    Dimension wd, ht;
    Position x, y;

    xfwfBoardClassRec.core_class.realize(self, mask, attributes);
    ((XfwfIconWidgetClass)self->core.widget_class)->xfwfCommon_class.compute_inside(self, &x, &y, &wd, &ht);
    XShapeCombineMask(XtDisplay(self), XtWindow(self), ShapeBounding, x, y,
		      ((XfwfIconWidget)self)->xfwfIcon.image ? ((XfwfIconWidget)self)->xfwfIcon.image->mask : None, ShapeSet);
}
#line 43 "Icon.w"
